[GeoMechanicsApplication] Add new element class for heat

markelov208 opened the pull request:
**📝 Description**
A brief description of the PR.

Please mark the PR with appropriate tags: 
- added transient thermal element

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- cleaned and refactored transient_thermal_element.h/cpp file
- registered instantiated classes of transient_thermal_element in geo_mechanics_application.h/cpp
- added unit tests for all instantiated classes

Commit:
- added private functions: VerifyProperty, CheckDomainSize, CheckSolutionStepsData

- made mIsInitialised as private
- removed some comments
- cleaned a bit
- removed this->
- applied c-lang format

Commit:
- applied c-lang format

Commit:
empty files for compilation

Commit:
used Variable.Name()

Commit:
- change transient_thermal_element.hpp to *.h

- added mTransientThermalElement2D3N

Commit:
- registered the rest of thermal elements

- test for 2D3N

Commit:
added tests for other thermal elements

Commit:
Merge branch 'master' into geo/11734-add-new-element-class-for-heat

Commit:
merged main into this branch

Commit:
- added mpCUrrectProcessInfo which is initialised via Initialize

- added check on mIsInitialised in CalculateConductivityMatrix
- added GeoThermalDispersionLaw geo in CalculateConductivityMatrix

Commit:
changed int => unsigned int

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,155 @@
+
+#include "custom_utilities/element_utilities.hpp"
+#include "geo_mechanics_application_variables.h"
+#include "includes/serializer.h"
```
I was a bit surprised to not see `#include "includes/element.h" in this list`, since `Element` is the base class of `TransientThermalElement`. So I expect it to be included indirectly, which I don't really like. If the indirect includes change (and I believe they will), that may break the building of this class. I would suggest to add `#include "includes/element.h"` to this list.

avdg81 replied:
This has been fixed.

avdg81 committed:
Removed an unused constructor

avdg81 committed:
Some member functions did not need to be `virtual`

avdg81 committed:
Function parameters passed by value shouldn't be `const`

See "Con.1: By default, make objects immutable" of the C++ Core
Guidelines. It states that "Function parameters passed by value are
rarely mutated, but also rarely declared `const`. To avoid confusion and
lots of false positives, don’t enforce this rule for function
parameters."

avdg81 committed:
Extracted the water density from `ElementVariables`

avdg81 committed:
Extracted the heat capacity of water from `ElementVariables`

avdg81 committed:
Extracted the solid density from `ElementVariables`

avdg81 committed:
Extracted the heat capacity of solid from `ElementVariables`

avdg81 committed:
Extracted the porosity from `ElementVariables`

avdg81 committed:
Extracted the saturation from `ElementVariables`

As a result, member function `InitializeProperties` became redundant. It
has been removed.

avdg81 committed:
Inlined the calculation of the integration coefficient

avdg81 committed:
Changed the access specifiers for several member functions

- Since class `TransientThermalElement` doesn't have any subclasses,
  there is no point in using `protected` member functions. Made those
  `private`.
- Corrected the access specifier of member function
  `CalculateLocalSystem`, which is `public` in the base class
  (`Element`), but its override was `private`. Now the override is
  `public` as well.

avdg81 committed:
Inlined the member function that returned the number of DOFs

avdg81 committed:
Inlined field `detJ` of the `ElementVariables`

avdg81 committed:
Inlined the calculation of the conductivity vector

Also removed the redundant field for the conductivity vector from
`ElementVariables`.

avdg81 committed:
Merged recent changes by Mohamed

The applied changes have been taken from commit `22e2e18`.

avdg81 committed:
Inlined the calculation of the capacity vector

Also removed the redundant field for the capacity vector from
`ElementVariables`.

avdg81 committed:
Eliminated two flags that were always set to `true`

avdg81 committed:
Removed the now redundant member `CalculateAndAddConductivityMatrix`

avdg81 committed:
Removed the now redundant member `CalculateAndAddCapacityMatrix`

avdg81 committed:
Removed an unused function parameter

avdg81 committed:
Eliminated field `ConstitutiveMatrix` of the `ElementVariables`

A local variable suffices.

avdg81 committed:
Removed an unused field from `ElementVariables`

avdg81 committed:
Inlined a member function

avdg81 committed:
Inlined a member function

avdg81 committed:
Inlined a member function

avdg81 committed:
Inlined a member function

avdg81 committed:
Inlined a member function

avdg81 committed:
Simplified some code

avdg81 committed:
Removed field `N` from the `ElementVariables`

It has been replaced by using a local variable.

avdg81 committed:
Removed field `NContainer` from the `ElementVariables`

It has been replaced by using a local variable.

avdg81 committed:
Removed field `GradNT` from the `ElementVariables`

It has been replaced by using a local variable.

avdg81 committed:
Removed a redundant flag

Considering the value of a pointer will do in this case.

avdg81 committed:
No longer store a pointer to the current process info

Instead, use the reference to the current process info object that is
passed to the member function. As a result, we no longer need the
override for `Initialize`.

avdg81 committed:
Extracted a member function that calculates the integration coefficients

avdg81 committed:
Eliminated field `ConductivityMatrix` from `ElementVariables`

Member function `CalculateConductivityMatrix` now calculates the
conductivity matrix of the element rather than the contribution of a
single integration point to the conductivity matrix.

avdg81 committed:
Eliminated field `CapacityMatrix` from `ElementVariables`

Member function `CalculateCapacityMatrix` now calculates the capacity
matrix of the element rather than the contribution of a single
integration point to the capacity matrix.

Due to the above changes, also field `IntegrationCoefficient` of the
`ElementVariables` was no longer used. It has been removed.

avdg81 committed:
Eliminated fields `detJContainer` and `DN_DXContainer`

These are now assigned to local variables and then passed via parameters
to the member functions that need this data.

avdg81 committed:
Eliminated fields `detJContainer` and `DN_DXContainer`

These are now assigned to local variables and then passed via parameters
to the member functions that need this data.

avdg81 committed:
Inlined a member function

avdg81 committed:
Extracted field `TemperatureVector` from `ElementVariables`

Added a function that builds up the temperature vector.

avdg81 committed:
Extracted field `DtTemperatureVector` from `ElementVariables`

Added a function that builds up the dt temperature vector. Since
`ElementVariables` had become empty by now, it has been removed as well.
Also the function that initialized the nodal temperature variables was
no longer used and thus has been removed.

avdg81 committed:
Removed an empty statement

avdg81 committed:
Merged two member functions that were very similar

avdg81 committed:
Removed two needless resize operations

avdg81 committed:
Inlined member function `CalculateAll`

avdg81 committed:
Apply fall-through when two cases must have the same effect

avdg81 committed:
Extracted a member function

The new member function adds contributions to the left hand side matrix.

avdg81 committed:
Extracted a member function

The new member function adds contributions to the right hand side
vector.

avdg81 committed:
Moved some implementations to the header file

avdg81 committed:
Removed two redundant local variables

Resizing the vector that contains the shape function's gradients is not
necessary, since the function that populates the vector also resizes it
when needed.

avdg81 committed:
Moved some implementations to the header file

Also made some stylistic improvements.

avdg81 committed:
Moved some implementations to the header file

avdg81 committed:
Moved the destructor's implementation to the header file

Also made some stylistic improvements.

avdg81 committed:
Moved a function implementation to the header file

Also simplified the function's body.

avdg81 committed:
Moved a function implementation to the header file

Also simplified the function's body.

avdg81 committed:
Moved a function implementation to the header file

Also simplified the function's body.

avdg81 committed:
Moved a function implementation to the header file

Also renamed its function parameter and a local variable.

avdg81 committed:
Moved a function implementation to the header file

Other changes include:
- Renamed a local variable.
- Removed an unneeded `#include`.

avdg81 committed:
Moved a function implementation to the header file

Also simplified the function's body.

avdg81 committed:
Moved a function implementation to the header file

Other changes include:
- Make sure that the access specifier of the overridden member function
  matches the one of the base class (i.e. `public`).
- Removed an alias.

avdg81 committed:
Moved a function implementation to the header file

Also simplified the function's body.

avdg81 committed:
Moved a function implementation to the header file

Also simplified the function's body.

avdg81 committed:
Refactored a member function

avdg81 committed:
Added a comprehensive check for solution steps data

The new function checks for all of the element's nodes whether they have
solution steps data for the given variable. Also make better use of
`const`.

avdg81 committed:
Extracted a member function that checks the degrees of freedom

Also reordered a few statements after a discussion with Wijtze Pieter.

avdg81 committed:
Eliminated two naked `new`s

In these cases, rather use `make_intrusive`.

avdg81 committed:
Extracted a function that checks all relevant properties

Also use `if constexpr` when possible (to avoid runtime branching).

avdg81 committed:
Renamed a member function

avdg81 committed:
Added a function that checks the Z coordinate in 2D models

avdg81 committed:
Moved a function's implementation to the header file

Also removed a redundant parameter name (it was not used).

avdg81 committed:
Extended the switch for the integration method

All expected numbers of nodes are now considered. The `default` now
throws an exception.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,32 @@
+
+namespace Kratos {
+
+template class TransientThermalElement<2, 3>;
```
It's not completely clear to me how this works. We specify certain instances of the template?

avdg81 replied:
This technique is called "template instantiation" and it is used for all (or nearly all) elements in the `GeoMechanicsApplication`. Basically, it forces the compiler to instantiate the class template given a set of template parameters. I have personally little experience with this technique (it was already used in this code base before I even joined the Kratos team). Although it may be useful here, we may not use it in the most optimal way. For instance, see this question on SO: https://stackoverflow.com/questions/2351148/explicit-template-instantiation-when-is-it-used. My suggestion is to investigate this in more detail later (e.g. by consulting some reference books), since it is a good question.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+        }
+    }
+
+    void CheckDomainSize() const
```
For a template class, is it also not possible to have these kinds of functions in the .cpp, which don't use the template arguments?

avdg81 replied:
I believe this question is partially related to your previous question. In general, member function definitions of class templates go in the header file. But there are situations where you may want to move them to a separate implementation file, as suggested by the question on SO: https://stackoverflow.com/questions/2351148/explicit-template-instantiation-when-is-it-used. Long story short: let's investigate the template instantiation mechanism in more detail first, and then we can decide about the best way forward. Do you agree?

rfaasse replied:
Agreed, let's keep it in mind for improvement possibilities

avdg81 committed:
A few more minor improvements

- Use a more logical `#include`.
- A few stylistic improvements.
- Simplified assembling the matrices. As a result, it is no longer
  needed to initialize the resulting matrices using zero matrices.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+
+    ~TransientThermalElement() override = default;
+    TransientThermalElement(const TransientThermalElement&) = delete;
+    TransientThermalElement& operator=(const TransientThermalElement&) = delete;
```
Not sure, but SQ might complain about the move assignment operator and move constructor (rule of 5).

avdg81 replied:
Good point. I have checked the copy operations of the base class (`Element`) and I've found those are actually implemented. As such, it is unclear to my why the copy operations are `delete`d in this derived class in the first place. Considering this, I would propose to apply the rule of zero, and let the compiler generate the needed special member functions.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+    {
+        KRATOS_TRY
+
+        rElementalDofList.resize(TNumNodes);
```
If the size is different, should we just continue, or do we want to log a warning/throw an error.

avdg81 replied:
I believe we should just continue. The point is that this member function returns its result through the argument list (`rElementalDofList`), and not using its return type. Although I would be in favor of changing the function's signature (to return the result using the return type rather than via an "out" parameter), this is a bit involved, since this member function is inherited from the Kratos Core class `Element`. As far as I can see, the size of `rElementalDofList` on entry has no meaning, so I believe we can safely resize it without further notice.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+    {
+        KRATOS_TRY
+
+        GeometryType::ShapeFunctionsGradientsType DN_DXContainer;
```
The name here does not adhere to the kratos style guide (same for detJContainer in the next line)

avdg81 replied:
Thanks for pointing out these inconsistencies. Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+    TransientThermalElement(const TransientThermalElement&) = delete;
+    TransientThermalElement& operator=(const TransientThermalElement&) = delete;
+
+    Element::Pointer Create(IndexType               NewId,
```
I thought all ::Pointers were shared pointer types, is it different for elements?

avdg81 replied:
Apparently, class `Element` uses macro `KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION`, which results in `Element::Pointer` being an alias for `Kratos::intrusive_ptr<Element>`. Perhaps at some point the `intrusive_ptr` class template may be replaced by a smart pointer from the STL? However, that is a discussion to be held with the Kratos Technical Committee. As such, it is out of scope for this PR.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+        KRATOS_CATCH("")
+
+        return 0;
+    }
```
In my opinion it would be nicer if the functions that are called here are placed directly under the check function

avdg81 replied:
Fair point. Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+        return result;
+    }
+
+    void CheckProperty(const Kratos::Variable<double>& rVariable) const
```
I would swap the order and place CheckProperty under CheckProperties

avdg81 replied:
Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,303 @@
+    void CheckProperty(const Kratos::Variable<double>& rVariable) const
+    {
+        KRATOS_ERROR_IF_NOT(GetProperties().Has(rVariable)) << rVariable.Name()
+                                                            << " does not exist in the material properties" << std::endl;
```
is 'material properties' in this case the right wording, maybe 'thermal element' is more descriptive?

avdg81 replied:
Agreed.

avdg81 committed:
Two minor improvements

- Corrected the level of indentation of a line.
- Removed an empty statement.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_transient_thermal_element.cpp:
```diff
@@ -0,0 +1,428 @@
+    }
+
+    const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+    // p_element->Check(r_current_process_info);
```
This is still commented, what do we want to do with this?

avdg81 replied:
I have removed it.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_transient_thermal_element.cpp:
```diff
@@ -0,0 +1,428 @@
+    rModelPart.CreateNewNode(1, 0.0, 0.0, 0.0);
+    rModelPart.CreateNewNode(2, 1.0, 0.0, 0.0);
+    rModelPart.CreateNewNode(3, 1.0, 1.0, 0.0);
+    std::vector<ModelPart::IndexType> elemNodes{1, 2, 3};
```
To adhere to the style guide: elemNodes -> elem_nodes (also in all other generate functions)

avdg81 replied:
Changed it to `node_ids`.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_transient_thermal_element.cpp:
```diff
@@ -0,0 +1,428 @@
+    // p_element->Check(r_current_process_info);
+    // Check requires a solution?
+
+    Element::DofsVectorType ElementalDofList;
```
To adhere to the style guide: ElementalDofList-> elemental_dof_list

avdg81 replied:
Done.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_transient_thermal_element.cpp:
```diff
@@ -0,0 +1,428 @@
+        ElementalDofList[i]->SetEquationId(i);
+    }
+
+    Element::EquationIdVectorType EquationIdVector;
```
To adhere to the style guide: ElementalDofList->equation_id_vector

avdg81 replied:
Renamed to `equation_ids`.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_transient_thermal_element.cpp:
```diff
@@ -0,0 +1,428 @@
+
+    // Check the EquationIdVector values
+    for (unsigned int i = 0; i < EquationIdVector.size(); i++) {
+        KRATOS_EXPECT_TRUE(EquationIdVector[i] == i);
```
We are missing some more asserts and checks on the element, but since we want to start testing with the integration tests as soon as possible, that could be a separate PR

avdg81 replied:
Agreed. I will create a new issue for this one.

rfaasse reviewed (commented):
The code looks very clean, I had no issues in reading and understanding the functionality. Thanks for all the effort @markelov208 @mnabideltares and @avdg81. I have some comments, but they are mainly related to order, formatting etc.

avdg81 committed:
A few minor improvements

- Removed a commented out line of code.
- Removed two redundant comments.
- Prefer to use `KRATOS_EXPECT_EQ` to compare two values rather than
  `KRATOS_EXPECT_TRUE` which receives a comparison expression.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/11734-add-new-element-class-for-heat

avdg81 committed:
Fixed a function call after merging changes from `master`

The function signature was modified. Consequently, the function call was
no longer valid (the number of function parameters was reduced). That
has been fixed now.

avdg81 committed:
Processed review feedback

- Applied the rule of zero to class `TransientThermalElement`.
- Renamed several local variables to comply with the Kratos Style Guide.
- Reordered the member functions that perform all kinds of checks, such
  that their order is more logical (i.e. it follows the order in which
  they are called from `Check`).
- Improved an error message.

avdg81 committed:
A few more fixes

- Corrected a comment.
- Reordered the check functions once more: only the entry function
  `Check` should be `public`. The API of the underlying checks should be
  `private`, to not clutter the public interface of the class.

avdg81 committed:
Added the `Geo` prefix to the new thermal elements

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,299 @@
+    BoundedMatrix<double, TNumNodes, TNumNodes> CalculateCapacityMatrix(const Vector& rIntegrationCoefficients) const
+    {
+        const auto& r_properties = GetProperties();
+        const auto  cWater = r_properties[POROSITY] * r_properties[SATURATION] *
```
couple more inconsistencies in this function (c_water, c_solid, n_container, g_point, n)

avdg81 replied:
Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,299 @@
+                law.CalculateThermalDispersionMatrix(GetProperties(), rCurrentProcessInfo);
+
+        auto result = BoundedMatrix<double, TNumNodes, TNumNodes>{ZeroMatrix{TNumNodes, TNumNodes}};
+        for (unsigned int GPoint = 0; GPoint < GetGeometry().IntegrationPointsNumber(GetIntegrationMethod()); ++GPoint) {
```
GPoint->g_point

avdg81 replied:
Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -0,0 +1,299 @@
+        const auto& r_integration_points = GetGeometry().IntegrationPoints(GetIntegrationMethod());
+
+        auto result = Vector{r_integration_points.size()};
+        for (unsigned int GPoint = 0; GPoint < r_integration_points.size(); ++GPoint) {
```
GPoint -> g_point

avdg81 replied:
Done.

rfaasse reviewed (commented):
Just a final few style inconsistencies, otherwise, good to go!

avdg81 committed:
Renamed a few local variables

Now they are consistent with the Kratos Style Guide.

rfaasse reviewed (approved):
Looks good!
